src/traits.jl

```diff
@@ -129,15 +129,16 @@ eltype(::Type{Colorant{T,N}}) where {T,N}   = T
 
 eltype(c::Colorant) = eltype(typeof(c))
 
-@pure function _parameter_upper_bound(t::UnionAll, idx)
-    Base.rewrap_unionall((Base.unwrap_unionall(t)::DataType).parameters[idx], t)
-end
-
 # eltypes_supported(Colorant{T<:X}) -> X
 function eltypes_supported(::Type{C}) where {C<:Colorant}
     Cb = base_colorant_type(C)
     isconcretetype(C) && C === Cb && return eltype(C)
-    _parameter_upper_bound(Cb, 1)
+    _eltypes_supported(Cb)
+end
+@pure function _eltypes_supported(::Type{C}) where {C<:Colorant}
+    nameof(C) === :Colorant || return _eltypes_supported(supertype(C))
+    C.var.name === :T || return Any
```

Comment:
Depending on `C.var.name` seems a little risky, though surprisingly it works for now:
```julia
julia> abstract type Container{T} end

julia> foo(::Type{C}) where C<:Container = C.var.name
foo (generic function with 1 method)

julia> foo(Container)
:T

julia> foo(Container{U} where U)
:T

julia> C = Container{U} where U
Container{U} where U

julia> C.var.name
:U
```

Reply:
Yes, this is a makeshift solution. I'm looking for a better way this past week.

Reply:
I changed the code to use `_parameter_upper_bound` again. However, the new implementation makes sure the input is a `Colorant`, so it is safer and more reasonable than the current master.

```julia
julia> ColorTypes.eltypes_supported(TransparentColor)
Color # current master
Any   # this PR

julia> ColorTypes.eltypes_supported(TransparentColor{RGB{T},T,4} where T) # analysis of constraints is still loose.
Any
```